sdk/state/update.go

```diff
@@ -0,0 +1,127 @@
+package state
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/stellar/experimental-payment-channels/sdk/txbuild"
+	"github.com/stellar/go/xdr"
+)
+
+type Payment struct {
+	IterationNumber       int64
+	AmountToInitiator     int64
+	AmountToResponder     int64
+	CloseSignatures       []xdr.DecoratedSignature
+	DeclarationSignatures []xdr.DecoratedSignature
+}
+
+// TODO - validate inputs? (eg. no negative amounts)
+// TODO - payments to be in Amount struct
+// initiator will only call this
+func (c *Channel) NewPayment(payToInitiator int64, payToResponder int64) (*Payment, error) {
+	newBalance := c.Balance + payToResponder - payToInitiator
+	amountToInitiator := int64(0)
+	amountToResponder := int64(0)
+	if newBalance > 0 {
+		amountToResponder = newBalance
+	} else {
+		amountToInitiator = newBalance * -1
+	}
+	txC, err := txbuild.Close(txbuild.CloseParams{
+		ObservationPeriodTime:      c.observationPeriodTime,
+		ObservationPeriodLedgerGap: c.observationPeriodLedgerGap,
+		InitiatorSigner:            c.localSigner.FromAddress(),
+		ResponderSigner:            c.remoteSigner,
+		InitiatorEscrow:            c.localEscrowAccount.Address,
+		ResponderEscrow:            c.remoteEscrowAccount.Address,
+		StartSequence:              c.startingSequence,
+		IterationNumber:            c.iterationNumber,
+		AmountToInitiator:          amountToInitiator,
+		AmountToResponder:          amountToResponder,
+	})
+	if err != nil {
+		return nil, err
+	}
+	txC, err = txC.Sign(c.networkPassphrase, c.localSigner)
+	if err != nil {
+		return nil, err
+	}
+	c.Balance = newBalance
+	return &Payment{
+		AmountToInitiator: payToInitiator,
+		AmountToResponder: payToResponder,
+		CloseSignatures:   txC.Signatures(),
+	}, nil
+}
+
+func (c *Channel) ConfirmPayment(p *Payment) (*Payment, error) {
+	if !c.initiator {
+		newBalance := c.Balance + p.AmountToResponder - p.AmountToInitiator
+		// TODO - better var names to differentiate from C_i fields?
+		amountToInitiator := int64(0)
+		amountToResponder := int64(0)
+		if newBalance > 0 {
+			amountToResponder = newBalance
+		} else {
+			amountToInitiator = newBalance * -1
+		}
+		txC, err := txbuild.Close(txbuild.CloseParams{
+			ObservationPeriodTime:      c.observationPeriodTime,
+			ObservationPeriodLedgerGap: c.observationPeriodLedgerGap,
+			InitiatorSigner:            c.remoteSigner,
+			ResponderSigner:            c.localSigner.FromAddress(),
+			InitiatorEscrow:            c.remoteEscrowAccount.Address,
+			ResponderEscrow:            c.localEscrowAccount.Address,
+			StartSequence:              c.startingSequence,
+			IterationNumber:            c.iterationNumber,
+			AmountToInitiator:          amountToInitiator,
+			AmountToResponder:          amountToResponder,
+		})
+		if err != nil {
+			return nil, err
+		}
+		if err := c.verifySigned(txC, p.CloseSignatures, c.remoteSigner); err != nil {
+			return nil, errors.Wrap(err, "the signed transaction may have different data")
```

Comment:
Instead of using errors.Wrap, can you use `fmt.Errorf` with `%w` that will wrap?

```suggestion
			return nil, fmt.Errorf("the signed transaction may have different data: %w", err)
```

The pkg/errors package has some big tradeoffs and idiosyncrasies, and I saw the author stated a while ago that they don't use it themselves. If you want to go down a rabbit hole reading about problems the package causes, see these: https://github.com/pkg/errors/issues/236, https://github.com/pkg/errors/issues/228, https://github.com/pkg/errors/issues/174, https://github.com/pkg/errors/issues/90.

I know we use pkg/errors in the stellar/go repo, but we should do everything we can to reduce where we introduce it now that the Go stdlib has the same capabilities built in since Go 1.13 without the problems linked above.

Reply:
TIL!
https://github.com/stellar/experimental-payment-channels/pull/55/commits/df02ced4c1911feeacc251acfb046fc1e2d03d19